File: src/design_patterns_with_java/clj/chapter02/file_io.clj
(ns design-patterns-with-java.clj.chapter02.file-io
  (:import
   (java.io
    FileInputStream
    FileOutputStream)
   (java.util
    Properties)))

(defn file-properties []
  (let [properties (Properties.)]
    {:read-from-file
     (fn [filename]
       (.. properties (load (FileInputStream. filename))))
     :write-to-file
     (fn [filename]
       (.. properties (store (FileOutputStream. filename)
                             "written by FileProperties")))
     :set-value
     (fn [key value]
       (.. properties (setProperty key value)))
     :get-value
     (fn [key]
       (.. properties (getProperty key)))}))

(defn -main []
  ;; map of closures
  (let [{:keys [read-from-file set-value write-to-file]}
        (file-properties)]
    (read-from-file "file.txt")
    (set-value "month" "4")
    (set-value "day" "21")
    (write-to-file "newfile1.txt")
    (set-value "year" "2004")
    (write-to-file "newfile2.txt")))

File: src/main/scala/design_patterns_with_java/scala/chapter02/FileIO.scala
package design_patterns_with_java.scala.chapter02

import java.io.{ FileInputStream, FileOutputStream }
import java.util.Properties

trait FileIO:
  def readFromFile(filename: String): Unit
  def writeToFile(filename: String): Unit
  def setValue(key: String, value: String): Unit
  def getValue(key: String): String

class FileProperties extends FileIO:
  private val properites = new Properties

  def readFromFile(filename: String): Unit =
    properites.load(new FileInputStream(filename))

  def writeToFile(filename: String): Unit =
    properites.store(new FileOutputStream(filename), "written by FileProperties")

  def setValue(key: String, value: String): Unit =
    properites.setProperty(key, value)

  def getValue(key: String): String =
    properites.getProperty(key)

@main
def fileIOExample() =
  // trait implementation by delegation
  val f: FileIO = new FileProperties
  f.readFromFile("file.txt")
  f.setValue("month", "4")
  f.setValue("day", "21")
  f.writeToFile("newfile1.txt")
  f.setValue("year", "2004")
  f.writeToFile("newfile2.txt")
